diff --git a/moderation_note.module b/moderation_note.module
index 9b371b9..e2c80e4 100644
--- a/moderation_note.module
+++ b/moderation_note.module
@@ -150,8 +150,14 @@ function moderation_note_preprocess_field(&$variables) {
     return;
   }
 
-  // Check the field type - we only support text fields at this time.
-  $supported_types = ['string', 'text', 'text_long', 'text_with_summary'];
+  // We only support text and entity reference revision fields at this time.
+  $supported_types = [
+    'entity_reference_revisions',
+    'string',
+    'text',
+    'text_long',
+    'text_with_summary',
+  ];
   if (!in_array($field_definition->getType(), $supported_types, TRUE)) {
     return;
   }
@@ -223,6 +229,12 @@ function _moderation_note_attach_field_notes(array &$variables) {
  *   TRUE if the current user can access the Entity, FALSE otherwise.
  */
 function _moderation_note_access(EntityInterface $entity) {
+  // If this entity is being referenced with entity reference revisions, it
+  // should not be notated individually.
+  if (isset($entity->_referringItem) && is_a($entity->_referringItem, '\Drupal\entity_reference_revisions\Plugin\Field\FieldType\EntityReferenceRevisionsItem')) {
+    return FALSE;
+  }
+
   /** @var \Drupal\content_moderation\ModerationInformation $moderation_information */
   $moderation_information = \Drupal::service('content_moderation.moderation_information');
 
