diff --git a/moderation_note.module b/moderation_note.module
index 08ac5bb..4bf7cfa 100644
--- a/moderation_note.module
+++ b/moderation_note.module
@@ -142,9 +142,6 @@ function moderation_note_preprocess_field(&$variables) {
   $element = $variables['element'];
   /** @var \Drupal\Core\Entity\EntityInterface $entity */
   $entity = $element['#object'];
-  /** @var \Drupal\Core\Field\FieldItemList $field_list */
-  $field_list = $element['#items'];
-  $field_definition = $field_list->getFieldDefinition();
 
   if (!_moderation_note_access($entity)) {
     return;
@@ -158,7 +155,8 @@ function moderation_note_preprocess_field(&$variables) {
     'text_long',
     'text_with_summary',
   ];
-  if (!in_array($field_definition->getType(), $supported_types, TRUE)) {
+  $field_type = $variables['field_type'] ? $variables['field_type'] : NULL;
+  if (!in_array($field_type, $supported_types, TRUE)) {
     return;
   }
 
