diff --git a/echo.install b/echo.install
index eeae389..1005ba2 100644
--- a/echo.install
+++ b/echo.install
@@ -32,12 +32,12 @@ function echo_requirements($phase) {
     $cached = cache_get($key);
     $pass = ($cached->data === $key);
   }
-  return array(
-    'echo' => array(
+  return [
+    'echo' => [
       'title' => t('Cache system'),
-      'description' => t('The <a href="!echo">%echo</a> module requires a working cache system.', array('!echo' => 'http://drupal.org/project/echo', '%echo' => 'Echo')),
+      'description' => t('The <a href="!echo">%echo</a> module requires a working cache system.', ['!echo' => 'http://drupal.org/project/echo', '%echo' => 'Echo']),
       'value' => $pass ? t('Working') : t('Not working'),
       'severity' => $pass ? REQUIREMENT_OK : REQUIREMENT_ERROR,
-    ),
-  );
+    ],
+  ];
 }
diff --git a/echo.module b/echo.module
index 3d4fd94..1225e9a 100644
--- a/echo.module
+++ b/echo.module
@@ -26,7 +26,7 @@
  *   A string containing the fully-themed html page.
  */
 function echo_themed_page($title, $content, $theme) {
-  $url = url('echo', array('absolute' => TRUE));
+  $url = url('echo', ['absolute' => TRUE]);
   // Store a hash of the arguments in the cache, which will be checked by
   // _echo_access() to ensure that the request originated from this function
   // and not from an external source.
@@ -34,13 +34,13 @@ function echo_themed_page($title, $content, $theme) {
   // Thirty seconds ought to be enough for anyone.
   $expiration = REQUEST_TIME + max(ini_get('max_execution_time'), 30);
   cache('cache')->set($key, $key, $expiration);
-  $options = array(
+  $options = [
     'method' => 'POST',
     'data' => 'title=' . rawurlencode($title)
     . '&content=' . rawurlencode($content)
     . '&theme=' . rawurlencode($theme),
-    'headers' => array('Content-Type' => 'application/x-www-form-urlencoded'),
-  );
+    'headers' => ['Content-Type' => 'application/x-www-form-urlencoded'],
+  ];
   $return = '';
   // Turn off maintenance mode so that anonymous page views work.
   if ($maintenance_mode = variable_get('maintenance_mode', 0)) {
@@ -61,18 +61,18 @@ function echo_themed_page($title, $content, $theme) {
  * Implements hook_menu().
  */
 function echo_menu() {
-  return array(
-    'echo' => array(
+  return [
+    'echo' => [
       'title callback' => '_echo_request',
-      'title arguments' => array('title'),
+      'title arguments' => ['title'],
       'page callback' => '_echo_request',
-      'page arguments' => array('content'),
+      'page arguments' => ['content'],
       'theme callback' => '_echo_request',
-      'theme arguments' => array('theme'),
+      'theme arguments' => ['theme'],
       'access callback' => '_echo_access',
       'type' => MENU_CALLBACK,
-    )
-  );
+    ]
+  ];
 }
 
 /**
