diff --git a/menu_block.module b/menu_block.module
index 8e424b7..51b52a9 100644
--- a/menu_block.module
+++ b/menu_block.module
@@ -1,5 +1,7 @@
 <?php
 
+use Drupal\Core\GeneratedLink;
+
 /**
  * Implements hook_theme_suggestions_HOOK() for "block".
  */
@@ -100,3 +102,21 @@ function menu_block_theme_suggestions_menu(array $variables) {
 
   return $suggestions;
 }
+
+/**
+ * Implements hook_preprocess_hook() for "block".
+ *
+ *   Some modules may be stripping the link off of the block title.
+ *   Replace the block label with the menu_block_configuration
+ *   label if a link exists.
+ */
+function menu_block_preprocess_block(&$variables) {
+
+  if (isset($variables['content']['#menu_block_configuration']['label'])) {
+    $config_label = $variables['content']['#menu_block_configuration']['label'];
+
+    if ($config_label instanceof GeneratedLink) {
+      $variables['label'] = $variables['content']['#menu_block_configuration']['label'];
+    }
+  }
+}
diff --git a/src/Plugin/Block/MenuBlock.php b/src/Plugin/Block/MenuBlock.php
index 2479c0e..5972074 100644
--- a/src/Plugin/Block/MenuBlock.php
+++ b/src/Plugin/Block/MenuBlock.php
@@ -10,6 +10,8 @@ use Drupal\Core\Entity\EntityTypeManagerInterface;
 use Drupal\Core\Menu\MenuActiveTrailInterface;
 use Drupal\Core\Menu\MenuLinkTreeInterface;
 use Drupal\Core\Menu\MenuTreeParameters;
+use Drupal\Core\Link;
+use Drupal\Core\Render\Markup;
 
 /**
  * Provides an extended Menu block.
@@ -138,6 +140,23 @@ class MenuBlock extends SystemMenuBlock {
       ],
     ];
 
+    $form['advanced']['label_link'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Link the title?'),
+      '#default_value' => $config['label_link'],
+      '#states' => [
+        'visible' => [
+          ':input[name="settings[label_type]"]' => [
+            ['value' => self::LABEL_ACTIVE_ITEM],
+            ['value' => self::LABEL_PARENT],
+            ['value' => self::LABEL_ROOT],
+            ['value' => self::LABEL_FIXED_PARENT ],
+            ['value' => 'initial_menu_item'],
+          ],
+        ],
+      ],
+    ];
+
     $form['style'] = [
       '#type' => 'details',
       '#title' => $this->t('HTML and style options'),
@@ -189,6 +208,7 @@ class MenuBlock extends SystemMenuBlock {
     $this->configuration['parent'] = $form_state->getValue('parent');
     $this->configuration['suggestion'] = $form_state->getValue('suggestion');
     $this->configuration['label_type'] = $form_state->getValue('label_type');
+    $this->configuration['label_link'] = $form_state->getValue('label_link');
   }
 
   /**
@@ -299,6 +319,7 @@ class MenuBlock extends SystemMenuBlock {
       'parent' => $this->getDerivativeId() . ':',
       'suggestion' => strtr($this->getDerivativeId(), '-', '_'),
       'label_type' => self::LABEL_BLOCK,
+      'label_link' => 0,
     ];
   }
 
@@ -435,10 +456,17 @@ class MenuBlock extends SystemMenuBlock {
    *   The menu item title.
    */
   protected function getLinkTitleFromLink($link_id) {
+    $config = $this->configuration;
     $parameters = new MenuTreeParameters();
     $menu = $this->menuTree->load($this->getDerivativeId(), $parameters);
     if ($link = $this->findLinkInTree($menu, $link_id)) {
-      return $link->link->getTitle();
+      if ($config['label_link']) {
+        $block_link = Link::fromTextAndUrl($link->link->getTitle(), $link->link->getUrlObject())->toString();
+        return Markup::create($block_link);
+      }
+      else {
+        return $link->link->getTitle();
+      }
     }
   }
 
