diff --git a/core/modules/block/block.module b/core/modules/block/block.module
index ec7e0ed7d8..1a281649d0 100644
--- a/core/modules/block/block.module
+++ b/core/modules/block/block.module
@@ -190,6 +190,8 @@ function block_theme_suggestions_block(array $variables) {
     $suggestions[] = 'block__' . $variables['elements']['#id'];
   }
 
+  // Remove duplicate suggestions.
+  $suggestions = array_keys(array_count_values($suggestions));
   return $suggestions;
 }
 
